File: behavioral_biometrics_nn/feature_extractor.py
"""Telemetry feature extraction.

Implements the 32-feature vector defined in ``design_choices.txt`` (FILE 2).
Features are computed per sampling window (1 s) over raw pointer events.

Input data format (Balabit-style session CSV)::

    record timestamp,client timestamp,button,state,x,y
    0.0,0.0,NoButton,Move,1043,410

Note on keyboard features (#21-26): the session files in this repository are
mouse-only (observed ``button`` values are NoButton/Left/Right/Scroll), so those
six slots evaluate to 0.0 here. The slots are kept so the vector layout stays
identical to the one produced by the JS telemetry client, which does emit
``keydown``/``keyup``.
"""

from __future__ import annotations

import csv
import math
from dataclasses import dataclass
from pathlib import Path

import numpy as np

WINDOW_SECONDS = 1.0
MIN_MOVE_EVENTS = 4
PAUSE_THRESHOLD_S = 0.2
DOUBLE_CLICK_MAX_S = 1.0
RESAMPLE_HZ = 50.0
TREMOR_BAND_HZ = (8.0, 12.0)

MOVEMENT_STATES = frozenset({"Move", "Drag"})
POINTER_BUTTONS = frozenset({"NoButton", "Left", "Right", "Middle", "Scroll"})
CORRECTION_KEYS = frozenset({"Backspace", "Delete"})

FEATURE_NAMES = (
    # 1. Mouse kinematics
    "v_mean", "v_max", "v_std",
    "a_mean", "a_max", "a_std",
    "j_mean", "j_max",
    # 2. Geometry & trajectory
    "path_efficiency", "total_curvature", "mean_angular_velocity",
    "dir_changes_x", "dir_changes_y", "inflection_points",
    # 3. Mouse rhythm & timing
    "click_dwell_mean", "click_dwell_std", "pause_count", "pause_ratio",
    "double_click_latency", "hover_duration_mean",
    # 4. Keyboard dynamics
    "key_dwell_mean", "key_dwell_std", "flight_time_mean", "flight_time_std",
    "typing_speed", "error_rate_ratio",
    # 5. Spectral & tremor dynamics
    "fft_peak_freq_x", "fft_peak_freq_y", "tremor_band_energy", "spectral_entropy",
    # 6. Context & interaction
    "scroll_speed_mean", "scroll_direction_changes",
)

N_FEATURES = len(FEATURE_NAMES)


@dataclass(frozen=True)
class Event:
    t: float
    button: str
    state: str
    x: float
    y: float


def read_session_events(session_path) -> list[Event]:
    """Parse a session CSV into chronologically ordered events."""
    path = Path(session_path)
    events: list[Event] = []
    with path.open("r", newline="", encoding="utf-8") as handle:
        reader = csv.DictReader(handle)
        if not reader.fieldnames or "record timestamp" not in reader.fieldnames:
            raise ValueError(f"Unexpected header in {path}: {reader.fieldnames}")
        for row in reader:
            try:
                events.append(
                    Event(
                        t=float(row["record timestamp"]),
                        button=row["button"],
                        state=row["state"],
                        x=float(row["x"]),
                        y=float(row["y"]),
                    )
                )
            except (TypeError, ValueError, KeyError):
                continue  # skip malformed rows
    events.sort(key=lambda e: e.t)
    return events


def segment_windows(
    events: list[Event],
    window_seconds: float = WINDOW_SECONDS,
    min_move_events: int = MIN_MOVE_EVENTS,
) -> list[list[Event]]:
    """Split an event stream into fixed-length windows.

    Windows containing fewer than ``min_move_events`` movement samples are
    dropped: they are the "invalid" vectors referenced by the enrolment step.
    """
    if not events:
        return []

    t0 = events[0].t
    buckets: dict[int, list[Event]] = {}
    for event in events:
        idx = int((event.t - t0) // window_seconds)
        buckets.setdefault(idx, []).append(event)

    windows = []
    for idx in sorted(buckets):
        window = buckets[idx]
        n_moves = sum(1 for e in window if e.state in MOVEMENT_STATES)
        if n_moves >= min_move_events:
            windows.append(window)
    return windows


def _movement_arrays(window: list[Event]):
    t, x, y = [], [], []
    for event in window:
        if event.state in MOVEMENT_STATES:
            t.append(event.t)
            x.append(event.x)
            y.append(event.y)
    return np.asarray(t, float), np.asarray(x, float), np.asarray(y, float)


def _kinematics(t, x, y) -> list[float]:
    """Features 1-8: velocity, acceleration and jerk statistics."""
    if t.size < 2:
        return [0.0] * 8

    dt = np.diff(t)
    step = np.hypot(np.diff(x), np.diff(y))
    valid = dt > 0
    if not np.any(valid):
        return [0.0] * 8

    v = step[valid] / dt[valid]
    t_v = t[1:][valid]
    v_stats = [float(v.mean()), float(v.max()), float(v.std())]

    a_stats = [0.0, 0.0, 0.0]
    j_stats = [0.0, 0.0]
    if v.size >= 2:
        dt_v = np.diff(t_v)
        ok = dt_v > 0
        if np.any(ok):
            a = np.diff(v)[ok] / dt_v[ok]
            t_a = t_v[1:][ok]
            a_stats = [float(a.mean()), float(a.max()), float(a.std())]
            if a.size >= 2:
                dt_a = np.diff(t_a)
                ok_a = dt_a > 0
                if np.any(ok_a):
                    j = np.diff(a)[ok_a] / dt_a[ok_a]
                    j_stats = [float(j.mean()), float(j.max())]

    return v_stats + a_stats + j_stats


def _geometry(t, x, y) -> list[float]:
    """Features 9-14: path shape and direction reversals."""
    if t.size < 2:
        return [0.0] * 6

    dx = np.diff(x)
    dy = np.diff(y)
    step = np.hypot(dx, dy)
    total_distance = float(step.sum())
    direct_distance = float(math.hypot(x[-1] - x[0], y[-1] - y[0]))
    path_efficiency = direct_distance / total_distance if total_distance > 0 else 0.0

    moved = step > 0
    total_curvature = 0.0
    mean_angular_velocity = 0.0
    if np.count_nonzero(moved) >= 2:
        theta = np.arctan2(dy[moved], dx[moved])
        d_theta = np.diff(theta)
        # wrap into [-pi, pi] so a 359 deg turn counts as -1 deg
        d_theta = (d_theta + np.pi) % (2 * np.pi) - np.pi
        total_curvature = float(np.abs(d_theta).sum())
        dt_theta = np.diff(t[1:][moved])
        ok = dt_theta > 0
        if np.any(ok):
            mean_angular_velocity = float(np.mean(np.abs(d_theta[ok]) / dt_theta[ok]))

    def _sign_changes(delta) -> int:
        signs = np.sign(delta)
        signs = signs[signs != 0]
        if signs.size < 2:
            return 0
        return int(np.count_nonzero(np.diff(signs) != 0))

    dir_changes_x = _sign_changes(dx)
    dir_changes_y = _sign_changes(dy)

    inflection_points = 0
    if dx.size >= 2:
        cross = dx[:-1] * dy[1:] - dy[:-1] * dx[1:]
        signs = np.sign(cross)
        signs = signs[signs != 0]
        if signs.size >= 2:
            inflection_points = int(np.count_nonzero(np.diff(signs) != 0))

    return [
        path_efficiency,
        total_curvature,
        mean_angular_velocity,
        float(dir_changes_x),
        float(dir_changes_y),
        float(inflection_points),
    ]


def _rhythm(window: list[Event], t_move) -> list[float]:
    """Features 15-20: click dwell, pauses, double-click latency, hover."""
    dwell_times: list[float] = []
    press_open: dict[str, float] = {}
    press_times: dict[str, list[float]] = {}

    for event in window:
        if event.button in ("NoButton", "Scroll"):
            continue
        if event.state == "Pressed":
            press_open[event.button] = event.t
            press_times.setdefault(event.button, []).append(event.t)
        elif event.state == "Released":
            start = press_open.pop(event.button, None)
            if start is not None and event.t >= start:
                dwell_times.append(event.t - start)

    click_dwell_mean = float(np.mean(dwell_times)) if dwell_times else 0.0
    click_dwell_std = float(np.std(dwell_times)) if len(dwell_times) > 1 else 0.0

    pause_count = 0
    pause_total = 0.0
    if t_move.size >= 2:
        gaps = np.diff(t_move)
        idle = gaps[gaps > PAUSE_THRESHOLD_S]
        pause_count = int(idle.size)
        pause_total = float(idle.sum())

    window_duration = window[-1].t - window[0].t if len(window) >= 2 else 0.0
    pause_ratio = (
        float(np.clip(pause_total / window_duration, 0.0, 1.0)) if window_duration > 0 else 0.0
    )

    latencies: list[float] = []
    for times in press_times.values():
        if len(times) < 2:
            continue
        deltas = np.diff(np.asarray(times, float))
        latencies.extend(float(d) for d in deltas if 0 < d <= DOUBLE_CLICK_MAX_S)
    double_click_latency = float(np.mean(latencies)) if latencies else 0.0

    hovers: list[float] = []
    if t_move.size:
        for event in window:
            if event.state != "Pressed" or event.button in ("NoButton", "Scroll"):
                continue
            before = t_move[t_move <= event.t]
            if before.size:
                hovers.append(float(event.t - before[-1]))
    hover_duration_mean = float(np.mean(hovers)) if hovers else 0.0

    return [
        click_dwell_mean,
        click_dwell_std,
        float(pause_count),
        pause_ratio,
        double_click_latency,
        hover_duration_mean,
    ]


def _keyboard(window: list[Event]) -> list[float]:
    """Features 21-26: keystroke dynamics (all zero for mouse-only sessions)."""
    key_events = [e for e in window if e.button not in POINTER_BUTTONS]
    if not key_events:
        return [0.0] * 6

    dwell: list[float] = []
    flight: list[float] = []
    open_keys: dict[str, float] = {}
    last_release: float | None = None
    press_count = 0
    correction_count = 0

    for event in key_events:
        if event.state == "Pressed":
            press_count += 1
            open_keys[event.button] = event.t
            if event.button in CORRECTION_KEYS:
                correction_count += 1
            if last_release is not None:
                flight.append(event.t - last_release)
        elif event.state == "Released":
            start = open_keys.pop(event.button, None)
            if start is not None and event.t >= start:
                dwell.append(event.t - start)
            last_release = event.t

    duration = key_events[-1].t - key_events[0].t
    typing_speed = press_count / duration if duration > 0 else 0.0
    error_rate = correction_count / press_count if press_count else 0.0

    return [
        float(np.mean(dwell)) if dwell else 0.0,
        float(np.std(dwell)) if len(dwell) > 1 else 0.0,
        float(np.mean(flight)) if flight else 0.0,
        float(np.std(flight)) if len(flight) > 1 else 0.0,
        float(typing_speed),
        float(error_rate),
    ]


def _spectral(t, x, y) -> list[float]:
    """Features 27-30: FFT of the resampled trajectory (tremor band, entropy)."""
    if t.size < 8:
        return [0.0] * 4

    # np.interp needs strictly increasing sample points
    unique_t, keep = np.unique(t, return_index=True)
    if unique_t.size < 8:
        return [0.0] * 4
    duration = float(unique_t[-1] - unique_t[0])
    if duration <= 0:
        return [0.0] * 4

    n = int(duration * RESAMPLE_HZ) + 1
    if n < 8:
        return [0.0] * 4

    grid = np.linspace(unique_t[0], unique_t[-1], n)
    xi = np.interp(grid, unique_t, x[keep])
    yi = np.interp(grid, unique_t, y[keep])

    taper = np.hanning(n)
    xi = (xi - xi.mean()) * taper
    yi = (yi - yi.mean()) * taper

    freqs = np.fft.rfftfreq(n, d=1.0 / RESAMPLE_HZ)
    if freqs.size < 2:
        return [0.0] * 4
    mag_x = np.abs(np.fft.rfft(xi))
    mag_y = np.abs(np.fft.rfft(yi))

    peak_x = float(freqs[int(np.argmax(mag_x[1:])) + 1])
    peak_y = float(freqs[int(np.argmax(mag_y[1:])) + 1])

    power = mag_x**2 + mag_y**2
    total_power = float(power.sum())
    band = (freqs >= TREMOR_BAND_HZ[0]) & (freqs <= TREMOR_BAND_HZ[1])
    tremor_energy = float(power[band].sum() / total_power) if total_power > 0 else 0.0

    spectral_entropy = 0.0
    if total_power > 0:
        p = power / total_power
        p = p[p > 0]
        spectral_entropy = float(-np.sum(p * np.log2(p)))

    return [peak_x, peak_y, tremor_energy, spectral_entropy]


def _scroll(window: list[Event]) -> list[float]:
    """Features 31-32: wheel speed and reversals.

    The dataset records wheel notches (``Scroll`` + ``Up``/``Down``) rather than
    pixel deltas, so ``scroll_speed_mean`` is expressed in notches per second.
    """
    times: list[float] = []
    deltas: list[float] = []
    for event in window:
        if event.button != "Scroll":
            continue
        if event.state == "Up":
            deltas.append(1.0)
        elif event.state == "Down":
            deltas.append(-1.0)
        else:
            continue
        times.append(event.t)

    if len(times) < 2:
        return [0.0, 0.0]

    dt = np.diff(np.asarray(times, float))
    ok = dt > 0
    speed = (
        float(np.mean(np.abs(np.asarray(deltas[1:], float))[ok] / dt[ok]))
        if np.any(ok)
        else 0.0
    )

    signs = np.sign(np.asarray(deltas, float))
    signs = signs[signs != 0]
    reversals = int(np.count_nonzero(np.diff(signs) != 0)) if signs.size >= 2 else 0

    return [speed, float(reversals)]


def extract_window_features(window: list[Event]) -> np.ndarray:
    """Compute the 32-feature vector for a single sampling window."""
    t, x, y = _movement_arrays(window)

    features: list[float] = []
    features += _kinematics(t, x, y)
    features += _geometry(t, x, y)
    features += _rhythm(window, t)
    features += _keyboard(window)
    features += _spectral(t, x, y)
    features += _scroll(window)

    if len(features) != N_FEATURES:
        raise AssertionError(f"expected {N_FEATURES} features, built {len(features)}")

    vector = np.asarray(features, dtype=np.float64)
    return np.nan_to_num(vector, nan=0.0, posinf=0.0, neginf=0.0).astype(np.float32)


def extract_session_windows(
    session_path,
    window_seconds: float = WINDOW_SECONDS,
    min_move_events: int = MIN_MOVE_EVENTS,
) -> np.ndarray:
    """Return an ``(n_windows, 32)`` matrix for one session file."""
    events = read_session_events(session_path)
    windows = segment_windows(events, window_seconds, min_move_events)
    if not windows:
        return np.empty((0, N_FEATURES), dtype=np.float32)
    return np.stack([extract_window_features(w) for w in windows])

File: demo-site/src/lib/biometrics/featureExtractor.ts
/**
 * TypeScript port of `behavioral_biometrics_nn/feature_extractor.py`'s
 * `extract_window_features`: the 32-feature vector computed per ~1 s window
 * of raw pointer events. Field names/units match what `BiometricsWidget.tsx`
 * already sends (`{t, x, y, button, state}`, with `button`/`state` using the
 * same string vocabulary as the Balabit-style training data: button in
 * "NoButton"/"Left"/"Right"/"Scroll", state in
 * "Move"/"Drag"/"Pressed"/"Released"/"Up"/"Down").
 *
 * Keyboard slots (features 21-26) are always zero here: the live browser
 * widget only ever puts mouse events in `events` (keystrokes go to the
 * separate `keys` array consumed by the keystroke module), exactly mirroring
 * how the Python `MouseBaseModule` never sees key events in its `events`
 * telemetry either.
 */

import { clip, mean, pstd, pyMod } from "./math";
import { hanning, interp, linspace, rfftFreqs, rfftMagnitude } from "./dft";
import type { MouseEvent2D } from "./types";

const PAUSE_THRESHOLD_S = 0.2;
const DOUBLE_CLICK_MAX_S = 1.0;
const RESAMPLE_HZ = 50.0;
const TREMOR_BAND_HZ: [number, number] = [8.0, 12.0];
const MOVEMENT_STATES = new Set(["Move", "Drag"]);

export const MIN_MOVE_EVENTS = 4;
export const N_FEATURES = 32;

function diff(xs: number[]): number[] {
  const out = new Array(Math.max(xs.length - 1, 0));
  for (let i = 1; i < xs.length; i++) out[i - 1] = xs[i] - xs[i - 1];
  return out;
}

function movementArrays(window: MouseEvent2D[]) {
  const t: number[] = [];
  const x: number[] = [];
  const y: number[] = [];
  for (const e of window) {
    if (MOVEMENT_STATES.has(e.state)) {
      t.push(e.t);
      x.push(e.x);
      y.push(e.y);
    }
  }
  return { t, x, y };
}

export function countMovementEvents(window: MouseEvent2D[]): number {
  let n = 0;
  for (const e of window) if (MOVEMENT_STATES.has(e.state)) n++;
  return n;
}

/** step/dt filtered to dt > 0, paired with the later timestamp of each pair. */
function velocityStage(t: number[], x: number[], y: number[]): { v: number[]; t: number[] } {
  const v: number[] = [];
  const tv: number[] = [];
  for (let i = 1; i < t.length; i++) {
    const dt = t[i] - t[i - 1];
    if (dt > 0) {
      v.push(Math.hypot(x[i] - x[i - 1], y[i] - y[i - 1]) / dt);
      tv.push(t[i]);
    }
  }
  return { v, t: tv };
}

/** Generic diff(value)/diff(t) filtered to dt > 0 - reused for acceleration
 * (from velocity) and jerk (from acceleration). */
function rateOfChange(t: number[], v: number[]): { v: number[]; t: number[] } {
  const out: number[] = [];
  const tout: number[] = [];
  for (let i = 1; i < v.length; i++) {
    const dt = t[i] - t[i - 1];
    if (dt > 0) {
      out.push((v[i] - v[i - 1]) / dt);
      tout.push(t[i]);
    }
  }
  return { v: out, t: tout };
}

function kinematics(t: number[], x: number[], y: number[]): number[] {
  if (t.length < 2) return new Array(8).fill(0);
  const { v, t: tv } = velocityStage(t, x, y);
  if (v.length === 0) return new Array(8).fill(0);
  const vStats = [mean(v), Math.max(...v), pstd(v)];

  let aStats = [0, 0, 0];
  let jStats = [0, 0];
  if (v.length >= 2) {
    const { v: a, t: ta } = rateOfChange(tv, v);
    if (a.length > 0) {
      aStats = [mean(a), Math.max(...a), pstd(a)];
      if (a.length >= 2) {
        const { v: j } = rateOfChange(ta, a);
        if (j.length > 0) jStats = [mean(j), Math.max(...j)];
      }
    }
  }
  return [...vStats, ...aStats, ...jStats];
}

function signChanges(delta: number[]): number {
  const signs = delta.map(Math.sign).filter((s) => s !== 0);
  if (signs.length < 2) return 0;
  let count = 0;
  for (let i = 1; i < signs.length; i++) if (signs[i] !== signs[i - 1]) count++;
  return count;
}

function geometry(t: number[], x: number[], y: number[]): number[] {
  if (t.length < 2) return new Array(6).fill(0);
  const dx = diff(x);
  const dy = diff(y);
  const step = dx.map((d, i) => Math.hypot(d, dy[i]));
  const totalDistance = step.reduce((a, b) => a + b, 0);
  const directDistance = Math.hypot(x[x.length - 1] - x[0], y[y.length - 1] - y[0]);
  const pathEfficiency = totalDistance > 0 ? directDistance / totalDistance : 0;

  const movedIdx: number[] = [];
  step.forEach((s, i) => {
    if (s > 0) movedIdx.push(i);
  });

  let totalCurvature = 0;
  let meanAngularVelocity = 0;
  if (movedIdx.length >= 2) {
    const theta = movedIdx.map((i) => Math.atan2(dy[i], dx[i]));
    const dTheta: number[] = [];
    for (let i = 1; i < theta.length; i++) {
      const raw = theta[i] - theta[i - 1];
      dTheta.push(pyMod(raw + Math.PI, 2 * Math.PI) - Math.PI);
    }
    totalCurvature = dTheta.reduce((a, b) => a + Math.abs(b), 0);
    const tMoved = movedIdx.map((i) => t[i + 1]);
    const dtTheta = diff(tMoved);
    const rates: number[] = [];
    dtTheta.forEach((d, i) => {
      if (d > 0) rates.push(Math.abs(dTheta[i]) / d);
    });
    if (rates.length > 0) meanAngularVelocity = mean(rates);
  }

  const dirChangesX = signChanges(dx);
  const dirChangesY = signChanges(dy);

  let inflectionPoints = 0;
  if (dx.length >= 2) {
    const cross: number[] = [];
    for (let i = 0; i < dx.length - 1; i++) cross.push(dx[i] * dy[i + 1] - dy[i] * dx[i + 1]);
    const signs = cross.map(Math.sign).filter((s) => s !== 0);
    if (signs.length >= 2) {
      let count = 0;
      for (let i = 1; i < signs.length; i++) if (signs[i] !== signs[i - 1]) count++;
      inflectionPoints = count;
    }
  }

  return [pathEfficiency, totalCurvature, meanAngularVelocity, dirChangesX, dirChangesY, inflectionPoints];
}

function rhythm(window: MouseEvent2D[], tMove: number[]): number[] {
  const dwellTimes: number[] = [];
  const pressOpen = new Map<string, number>();
  const pressTimes = new Map<string, number[]>();

  for (const e of window) {
    if (e.button === "NoButton" || e.button === "Scroll") continue;
    if (e.state === "Pressed") {
      pressOpen.set(e.button, e.t);
      const arr = pressTimes.get(e.button) ?? [];
      arr.push(e.t);
      pressTimes.set(e.button, arr);
    } else if (e.state === "Released") {
      const start = pressOpen.get(e.button);
      if (start !== undefined) {
        pressOpen.delete(e.button);
        if (e.t >= start) dwellTimes.push(e.t - start);
      }
    }
  }

  const clickDwellMean = dwellTimes.length ? mean(dwellTimes) : 0;
  const clickDwellStd = dwellTimes.length > 1 ? pstd(dwellTimes) : 0;

  let pauseCount = 0;
  let pauseTotal = 0;
  if (tMove.length >= 2) {
    for (const g of diff(tMove)) {
      if (g > PAUSE_THRESHOLD_S) {
        pauseCount++;
        pauseTotal += g;
      }
    }
  }

  const windowDuration = window.length >= 2 ? window[window.length - 1].t - window[0].t : 0;
  const pauseRatio = windowDuration > 0 ? clip(pauseTotal / windowDuration, 0, 1) : 0;

  const latencies: number[] = [];
  for (const times of pressTimes.values()) {
    if (times.length < 2) continue;
    for (let i = 1; i < times.length; i++) {
      const d = times[i] - times[i - 1];
      if (d > 0 && d <= DOUBLE_CLICK_MAX_S) latencies.push(d);
    }
  }
  const doubleClickLatency = latencies.length ? mean(latencies) : 0;

  const hovers: number[] = [];
  if (tMove.length) {
    for (const e of window) {
      if (e.state !== "Pressed" || e.button === "NoButton" || e.button === "Scroll") continue;
      let before = -Infinity;
      for (const tm of tMove) if (tm <= e.t && tm > before) before = tm;
      if (before !== -Infinity) hovers.push(e.t - before);
    }
  }
  const hoverDurationMean = hovers.length ? mean(hovers) : 0;

  return [clickDwellMean, clickDwellStd, pauseCount, pauseRatio, doubleClickLatency, hoverDurationMean];
}

function spectral(t: number[], x: number[], y: number[]): number[] {
  if (t.length < 8) return [0, 0, 0, 0];

  const uniqueT: number[] = [];
  const uniqueX: number[] = [];
  const uniqueY: number[] = [];
  for (let i = 0; i < t.length; i++) {
    if (i === 0 || t[i] !== t[i - 1]) {
      uniqueT.push(t[i]);
      uniqueX.push(x[i]);
      uniqueY.push(y[i]);
    }
  }
  if (uniqueT.length < 8) return [0, 0, 0, 0];
  const duration = uniqueT[uniqueT.length - 1] - uniqueT[0];
  if (duration <= 0) return [0, 0, 0, 0];

  const n = Math.floor(duration * RESAMPLE_HZ) + 1;
  if (n < 8) return [0, 0, 0, 0];

  const grid = linspace(uniqueT[0], uniqueT[uniqueT.length - 1], n);
  let xi = interp(grid, uniqueT, uniqueX);
  let yi = interp(grid, uniqueT, uniqueY);

  const taper = hanning(n);
  const xMean = mean(xi);
  const yMean = mean(yi);
  xi = xi.map((v, i) => (v - xMean) * taper[i]);
  yi = yi.map((v, i) => (v - yMean) * taper[i]);

  const freqs = rfftFreqs(n, 1 / RESAMPLE_HZ);
  if (freqs.length < 2) return [0, 0, 0, 0];
  const magX = rfftMagnitude(xi);
  const magY = rfftMagnitude(yi);

  const argmaxFrom1 = (mags: number[]) => {
    let best = 1;
    let bestVal = mags[1] ?? 0;
    for (let i = 2; i < mags.length; i++) {
      if (mags[i] > bestVal) {
        bestVal = mags[i];
        best = i;
      }
    }
    return best;
  };
  const peakX = freqs[argmaxFrom1(magX)];
  const peakY = freqs[argmaxFrom1(magY)];

  const power = magX.map((m, i) => m * m + magY[i] * magY[i]);
  const totalPower = power.reduce((a, b) => a + b, 0);

  let tremorEnergy = 0;
  if (totalPower > 0) {
    let bandSum = 0;
    power.forEach((p, i) => {
      const f = freqs[i];
      if (f >= TREMOR_BAND_HZ[0] && f <= TREMOR_BAND_HZ[1]) bandSum += p;
    });
    tremorEnergy = bandSum / totalPower;
  }

  let spectralEntropy = 0;
  if (totalPower > 0) {
    const p = power.map((v) => v / totalPower).filter((v) => v > 0);
    spectralEntropy = -p.reduce((a, v) => a + v * Math.log2(v), 0);
  }

  return [peakX, peakY, tremorEnergy, spectralEntropy];
}

function scrollFeatures(window: MouseEvent2D[]): number[] {
  const times: number[] = [];
  const deltas: number[] = [];
  for (const e of window) {
    if (e.button !== "Scroll") continue;
    if (e.state === "Up") deltas.push(1);
    else if (e.state === "Down") deltas.push(-1);
    else continue;
    times.push(e.t);
  }
  if (times.length < 2) return [0, 0];

  const dt = diff(times);
  const rates: number[] = [];
  for (let i = 0; i < dt.length; i++) if (dt[i] > 0) rates.push(Math.abs(deltas[i + 1]) / dt[i]);
  const speed = rates.length ? mean(rates) : 0;

  const signs = deltas.map(Math.sign).filter((s) => s !== 0);
  let reversals = 0;
  if (signs.length >= 2) for (let i = 1; i < signs.length; i++) if (signs[i] !== signs[i - 1]) reversals++;

  return [speed, reversals];
}

/** Compute the 32-feature vector for one ~1 s batch of raw mouse events. */
export function extractWindowFeatures(window: MouseEvent2D[]): number[] {
  const sorted = [...window].sort((a, b) => a.t - b.t);
  const { t, x, y } = movementArrays(sorted);

  const features = [
    ...kinematics(t, x, y),
    ...geometry(t, x, y),
    ...rhythm(sorted, t),
    0, 0, 0, 0, 0, 0, // keyboard slots 21-26: always zero, see module docstring
    ...spectral(t, x, y),
    ...scrollFeatures(sorted),
  ];

  if (features.length !== N_FEATURES) {
    throw new Error(`expected ${N_FEATURES} features, built ${features.length}`);
  }
  return features.map((v) => (Number.isFinite(v) ? v : 0));
}
